api/src/main/java/org/apache/iceberg/expressions/ManifestEvaluator.java

```diff
@@ -144,18 +143,37 @@ public Boolean or(Boolean leftResult, Boolean rightResult) {
     @Override
     public <T> Boolean isNaN(BoundReference<T> ref) {
       int pos = Accessors.toPosition(ref.accessor());
-      // containsNull encodes whether at least one partition value is null, lowerBound is null if
-      // all partition values are null.
-      if (stats.get(pos).containsNull() && stats.get(pos).lowerBound() == null) {
-        return ROWS_CANNOT_MATCH; // all values are null
+
+      if (stats.get(pos).containsNaN() != null && !stats.get(pos).containsNaN()) {
+        return ROWS_CANNOT_MATCH;
+      }
+
+      if (allValuesAreNull(stats.get(pos))) {
+        return ROWS_CANNOT_MATCH;
       }
 
       return ROWS_MIGHT_MATCH;
     }
 
+    private boolean allValuesAreNull(PartitionFieldSummary summary) {
+      // Before introducing containsNaN field, containsNull encodes whether at least one partition value is null,
+      // lowerBound is null if all partition values are null.
+      // After introducing containsNaN field, containsNaN must be false to ensure all values are null since bounds
+      // don't include NaN anymore.
+      return summary.containsNull() && summary.lowerBound() == null &&
+          (summary.containsNaN() == null || !summary.containsNaN());
```

Comment:
I understand that this is trying to retain the previous behavior, but if `containsNaN` is not known, it seems safer to assume there is a NaN value.

I think there is an argument for doing this: at one point, NaN wasn't handled separately and if `lower` and `upper` were null then there were no non-null values. A `NaN` would cause a bound to be non-null. The problem is that we must accept manifests that were written with the new rule (don't store NaN for `lower`), but that didn't set this. I'd prefer to go with the safer option.

Reply:
One more thing: if we go for the safer option, we only need to do that for float and double columns. All other types can't contain NaN.

Reply:
I think the change for excluding NaN in `lower`/`upper` and adding `containsNaN` both belong to this PR, so if a release contains this change, then it would either be (1) `NaN` is part of `lower`/`upper` and `containsNaN` is missing, or (2) `containsNaN` exists and `lower`/`upper` doesn't store `NaN`. But I guess people may implement their own manifest summary that already exclude `NaN` from bounds but no `containsNaN`, so we still want to handle this, and file level metrics could give more granular information so there isn't necessarily any performance penalty. I have updated this PR to check for existence of `containsNaN`, but please let me know if my understanding isn't correct!

Reply:
Yes, I think your summary is correct. Because this is a format change and we expect/want others to implement the format, we want to account for strange cases like that.